block/local/adapter_test.go

```diff
@@ -141,3 +144,79 @@ func TestLocalCopy(t *testing.T) {
 		t.Errorf("expected to read \"%s\" as written, got \"%s\"", contents, string(got))
 	}
 }
+
+func dumpPathTree(t testing.TB, root string) []string {
+	t.Helper()
+	tree := make([]string, 0)
+	err := filepath.Walk(root, func(path string, _ os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		p := strings.TrimPrefix(path, root)
+		tree = append(tree, p)
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("walking on '%s': %s", root, err)
+	}
+	return tree
+}
+
+func TestAdapter_Remove(t *testing.T) {
+	const content = "Content used for testing"
+	tests := []struct {
+		name              string
+		additionalObjects []string
+		path              string
+		wantErr           bool
+		wantTree          []string
+	}{
+		{
+			name:     "single",
+			path:     "README",
+			wantErr:  false,
+			wantTree: []string{""},
+		},
+
+		{
+			name:     "under folder",
+			path:     "src/tools.go",
+			wantErr:  false,
+			wantTree: []string{""},
+		},
+		{
+			name:     "under multiple folders",
+			path:     "a/b/c/d.txt",
+			wantErr:  false,
+			wantTree: []string{""},
+		},
+		{
+			name:              "file in the way",
+			path:              "a/b/c/d.txt",
+			additionalObjects: []string{"a/b/blocker.txt"},
+			wantErr:           false,
+			wantTree:          []string{"", "/test", "/test/a", "/test/a/b", "/test/a/b/blocker.txt"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// setup env
+			adp := makeAdapter(t)
+			envObjects := append(tt.additionalObjects, tt.path)
+			for _, o := range envObjects {
+				obj := makePointer(o)
+				testutil.MustDo(t, "Put", adp.Put(obj, 0, strings.NewReader(content), block.PutOpts{}))
+			}
+			// test Remove with remove empty folders
+			obj := makePointer(tt.path)
+			if err := adp.Remove(obj); (err != nil) != tt.wantErr {
+				t.Errorf("Remove() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			tree := dumpPathTree(t, adp.Path())
+			if diff := deep.Equal(tt.wantTree, tree); diff != nil {
```

Comment:
This comparison will be flaky AFAICT, depending even on properties of the underlying *filesystem*!  `filepath.Walk` doesn't guarantee order of returned objects.  Best to sort `tree` first.

Reply:
I forgot the sort!